milter-greylist.1
.\"
.\" $Id: milter-greylist.1,v 1.1 2004/02/21 17:46:22 manu Exp $
.\"
.\" Copyright (c) 2004 Emmanuel Dreyfus
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"        This product includes software developed by Emmanuel Dreyfus
.\"
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
.\" OF THE POSSIBILITY OF SUCH DAMAGE.
.\"

.Dd Feb 21, 2004
.Dt milter-greylist 8
.Os
.Sh NAME
.Nm milter-greylist
.Nd grey listing filter for sendmail
.Sh SYNOPSIS
.Nm
.Op Fl v
.Op Fl d Ar dumpfile
.Op Fl f Ar exceptfile
.Op Fl w Ar delay
.Fl p Ar socket
.Sh DESCRIPTION
.Nm 
is a mail filter for sendmail that implements grey listing,
a spam filtering technique proposed by Evan Harris.
.Pp
Grey listing works by assuming that contrarily to legitimate MTA, spam engines
will not retry sending their junk mail on a temporary error. The filter
will always temporarilly reject mail on a first attempt, and to 
accept it after some time has elapsed.
.Pp
If spammers ever try to resend rejected messages, we can assume they will 
not stay idle between the two sends. Odds are good that the spammer will 
send a mail to an honey pot address and get blacklisted in a distributed 
black list before the second attempt.
.Pp
Of course, the filter can be configured to not apply grey listing to some
hosts or networks. You can white list friendly SMTP servers, and you should
whitelist your own network, else your SMTP clients will have real trouble to 
send e-mail. Whitelisting localhost is also a must.
.Pp
.Nm
works with two files. 
.Pa except.conf
holds the white list of machines that will not suffer grey list filtering.
It is readen upon 
.Nm
startup and is left alone once started. The syntax of this file is simple:
it just consist of one pair of network address and CIDR netmask per line, 
separated by a / (slash). Here is an example:
.Dl 127.0.0.0/8
.Dl 192.168.3.0/24
.Pp
The second file is
.Pa greylist.db .
.Nm
will regularly dump its grey list database into this file, which is used
on startup to restore the previous grey list state. If the file does not
exists or is unreadable, 
.Nm
will start with an empty grey list. 
.Pp
The following options are available:
.Bl -tag -width flag
.It Fl d Ar dumpfile
Location of the dump file. Default is 
.Pa /etc/mail/except.conf .
.It Fl f Ar exceptfile
Location of the white list file. Default is
.Pa /var/milter-greylist/greylist.db .
.It Fl p Ar socket
Use 
.Ar socket
as the socket used by 
.Xr sendmail 8
to talk with
.Nm .
This option is mandatory.
.It Fl v
Enable debug output. 
.Nm 
will send messages (and debug output if it is given the
.Fl v
flag) to 
.Xr syslogd 8
with facility LOG_MAIL.
.It Fl w Ar delay
sets the minimum delay (in seconds) between the first attempt and the time
the message can be accepted. Default is 1800 seconds (30 minutes).
.El
.Sh AUTHOR
.An Emmanuel Dreyfus Aq manu@netbsd.org
.Sh SEE ALSO
.Xr sendmail 8 , 
.Xr syslogd 8 .
.Pp
Evan Harris's paper
.Pa http://projects.puremagic.com/greylisting


ChangeLog
4.0a3
	Add an urlcheck reply which is ignored: milterGreylistIgnore 
	Build fixes for Debian (Bernhard Schneider)
	Add a configure flag to not use --rpath (Bernhard Schneider)
	Cleanup stale PID file on startup (Bernhard Schneider)
	Fix memory leak with DNSRBL (Michael Fromme)
4.0a2
	For urlcheck answer, cope with a trailing line not terminated by CR
	Fix X-Greylist lossage (AIDA Shinra)
	Fix garbled log (Bernhard Schneider)
	Fix build on Tru64 (Attila Bruncsak)
4.0a1
	Fix again an accept-all-bug when built with SPF (AIDA Shinra)
	Fix macro handling in ACL (AIDA Shinra)
	%% syntax in format strings (AIDA Shinra)
	Clarified "%Xm" and "%Xh" semantics (AIDA Shinra)
	Fixed a memory leak at mlfi_eom (AIDA Shinra)
	Fix a bug where doing RCPT twice when blacklised succeed (AIDA Shinra)
3.1.8
	Allow filtering on the HELO string
	Back out previous SPF fix, it caused accept-all with SPF-less builds
3.1.7
	Fix an accept-all-bug when built with SPF (AIDA Shinra)
	Option to perform urlchecks in forked instance to avoid thread-unsafety
	Allow escaping of " in strings and / in regex
	Fix whitelisting using access.db (Georg Horn)
	Fix NULL pointer referencing in urlcheck code
	Fix body storage so that locators (^ and $) work for body regex
	Fix spurious warning about unknown whitelisting conditions
	Fix memory leaks when using urlchecks
	Makes urlcheck properties and values case insensitives
	Do not retain urlcheck properties if the urlcheck clause did not match
	Allow loading regex in lists of body clauses
	Option to clear urlcheck prop before handling a new recipient
	%D format string for getting the list of matching DNSRBL
	Avoid performing multiple DNSRBL checks for the same IP
	Allow reusing in the ACL of properties gathered from urlcheck
	Fix wrong display of %Xc %Xe and %Xh substitutions 
	Fix documentation: %Xh instead of %Xr
	stat should not report X-Greylist header is message was rejected
	Fix a documentation bug: sender e-mail is %f, not %s
3.1.6
	More format strings, for SMTP code, extended code, message, header
	Fix a crash when using regex without grouping ()
	Document the CVS location in README
3.1.5
	Fix double free when using content filtering
	ACL clauses can now be negated
	msg clauses in ACL can use format strings substitution
	Add an ACL clause to customize X-Greylist header
	Fix serious bugs in DNSRBL code (Jacques Beigbeder)
	Fix a display bug for netblocks
	Add a time clause to match against time sets
	GeoIP support, through the geoip clause
	Fix warning for Postfix build (Nerijus Baliunas)
	%g for substitution by regex back references
	%I for susbtitution by sender IP masked by a CIDR
	Allow specifying socket mode in config file
	Set default user to root for .spec file
	Do not drop root privs if we do not run as root
3.1.4
	Add support for switching to a given group (Ralf S. Engelschall)
	Add substitutions for sendmail macro and strftime in URL checks
	Add custom logs of milter-greylist actions, see stat in greylist.conf
	Fixes for running with Postfix (Nerijus Baliunas)
	Fix configure problem with DNSRBL on Linux (Andrew McGill)
	Document how to use milter-greylist with Postfix (Nerijus Baliunas)
	Update .spec for Postfix (Nerijus Baliunas)
	Optionnaly post the message body to an URL check at DATA stage
	Integrate SPF, SMTP AUTH and STARTTLS in ACL
	Add msgsize and rcptcount clauses to ACL
	Allow rcpt clause at DATA stage ACL
	Fix crashes when running with -D on some systems (John Thiltges)
	Do not sort the databases, it makes startup slow and buys nothing
3.1.3
	Fix various bugs (regex searches, lists matches, uninitialized memory)
3.1.2
	Fix timespamp on sparc64 (Gert Doering)
	Add the ability to query extrnal sources in ACL using URL 
	Add connexion pools for URL queries to enable persistent connexions
	Fix dependency in rc-bsd.sh: it's mail, not sendmail
	Add DATA-stage ACL
	Update to .spec file (Rudy Eschauzier)
	Allow header and body searches in DATA-stage ACL
	Allow CIDR match for DNSRBL
	Allow multiple macro, dnsrbl, urlcheck, body and header clauses in ACL
3.1.1
	Fix crashes during dump reloads (AIDA Shinra)
	Fix DoS in MX sync protocol (AIDA Shinra)
	Check for -lc_r before others, for FreeBSD (AIDA Shinra)
	Fix configure for libspf2 (AIDA Shinra)
	Fix FreeBSD build (AIDA Shinra)
	Avoid buffer overflow in DNSRBL code (AIDA Shinra)
	Fix build problem with newer BerkeleyDB (AIDA Shinra)
	Check if -lbind requires -lpthread, for Linux
	Correctly enable non blocking I/O fox MX sync (Attila Bruncsak)
	Pointer to DRAC documentation (Matthias Scheler)
	FreeBSD build fixes (Hajimu UMEMOTO)
	Remove the /tmp/access-list.debug for security sake (AIDA Shinra)
	Fix display bug in log messages (AIDA Shinra)
	Updated the list of broken MTA
	Fix MX sync between Tru64 and Linux (Attila Bruncsak)
	Silly build fix for platforms that lack vsyslog()
	Honour LINE_MAX for syslog (Attila Bruncsak)
	Warn about ignored ACL lines after acl default rule
	Fix various race conditions (AIDA Shinra)
	Fix big bugs in macro support (AIDA Shinra)
	Fix build warning on Tru64
	Build fix on Solaris
	Documentation fix
3.0a5
	Build fix on Tru64, fix MX sync (Attila Bruncsak)
	Fix various race conditions (AIDA Shinra)
3.0a4
	IPv6 and build fixes (Hajimu UMEMOTO)
3.0a3
	Fix --disable-drac and --disable-dnsrbl (Fabien Tassin)
	Add support to use Sendmail macros in the ACL
	Remove the need of LOG_PERROR, make sure no log is missed
	Make sure -c tells what happens and return an usable exit code
3.0a2
	Documentation fix about required sendmail macros
	Missing bit in rc-redhat.sh (Jobst Schmalenbach)
	Fix bugs that prevented IPv6 addresses resolution
	Add more log output for blackist and flushaddr (Fabien Tassin)
	Allow rejecting by tempfail on blacklist configuration (Fabien Tassin)
3.0a1
	Build fix
2.1.12
	IPv6 fixes again (Hajimu UMEMOTO)
	Build fix on FreeBSD
	per-ACL SMTP error code and messages
2.1.11
	IPv6 fixes (Hajimu UMEMOTO)
	flushaddr flag for access lists, which flush an IP from the database
2.1.10

milter-greylist.c
/* $Id: milter-greylist.c,v 1.50 2004/03/23 14:01:03 manu Exp $ */

/*
 * Copyright (c) 2004 Emmanuel Dreyfus
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *        This product includes software developed by Emmanuel Dreyfus
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,  
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"

#ifdef HAVE_SYS_CDEFS_H
#include <sys/cdefs.h>
#ifdef __RCSID  
__RCSID("$Id: milter-greylist.c,v 1.50 2004/03/23 14:01:03 manu Exp $");
#endif
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <syslog.h>
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <pwd.h>
#include <unistd.h>

/* On IRIX, <unistd.h> defines a EX_OK that clashes with <sysexits.h> */
#ifdef EX_OK
#undef EX_OK
#endif
#include <sysexits.h>

#if HAVE_GETOPT_H
#include <getopt.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>

#include <libmilter/mfapi.h>

#include "dump.h"
#include "except.h"
#include "conf.h"
#include "pending.h"
#include "sync.h"
#include "autowhite.h"
#include "milter-greylist.h"

int debug = 0;
int dont_fork = 0;
int quiet = 0;

static char *strncpy_rmsp(char *, char *, size_t);
static int humanized_atoi(char *);

struct smfiDesc smfilter =
{
	"greylist",	/* filter name */
	SMFI_VERSION,	/* version code */
	SMFIF_ADDHDRS,	/* flags */
	mlfi_connect,	/* connection info filter */
	NULL,		/* SMTP HELO command filter */
	mlfi_envfrom,	/* envelope sender filter */
	mlfi_envrcpt,	/* envelope recipient filter */
	NULL,		/* header filter */
	NULL,		/* end of header */
	NULL,		/* body block filter */
	mlfi_eom,	/* end of message */
	NULL,		/* message aborted */
	mlfi_close,	/* connection cleanup */
};

sfsistat
mlfi_connect(ctx, hostname, addr)
	SMFICTX *ctx;
	char *hostname;
	_SOCK_ADDR *addr;
{
	struct mlfi_priv *priv;
	struct sockaddr_in *addr_in;

	if ((priv = malloc(sizeof(*priv))) == NULL)
		return SMFIS_TEMPFAIL;	

	smfi_setpriv(ctx, priv);
	bzero((void *)priv, sizeof(*priv));
	priv->priv_whitelist = EXF_UNSET;

	addr_in = (struct sockaddr_in *)addr;
	if ((addr_in != NULL) && (addr_in->sin_family == AF_INET))
		priv->priv_addr.s_addr = addr_in->sin_addr.s_addr;

	return SMFIS_CONTINUE;
}

sfsistat
mlfi_envfrom(ctx, envfrom)
	SMFICTX *ctx;
	char **envfrom;
{
	struct mlfi_priv *priv;

	priv = (struct mlfi_priv *) smfi_getpriv(ctx);

	/*
	 * Strip spaces from the source address
	 */
	strncpy_rmsp(priv->priv_from, *envfrom, ADDRLEN);
	priv->priv_from[ADDRLEN] = '\0';

	return SMFIS_CONTINUE;
}

sfsistat
mlfi_envrcpt(ctx, envrcpt)
	SMFICTX *ctx;
	char **envrcpt;
{
	struct mlfi_priv *priv;
	time_t remaining;
	char hdr[HDRLEN + 1];
	char addrstr[IPADDRLEN + 1];
	char rcpt[ADDRLEN + 1];
	int h, mn, s;

	priv = (struct mlfi_priv *) smfi_getpriv(ctx);

	if ((priv->priv_queueid = smfi_getsymval(ctx, "{i}")) == NULL) {
		syslog(LOG_DEBUG, "smfi_getsymval faild for {i}: %s",
		    strerror(errno));
		priv->priv_queueid = "(unkown id)";
	}
	
	if (debug)
		syslog(LOG_DEBUG, "%s: addr = %s, from = %s, rcpt = %s", 
		    priv->priv_queueid, inet_ntoa(priv->priv_addr), 
		    priv->priv_from, *envrcpt);

	/*
	 * For multiple-recipients messages, if the sender IP or the
	 * sender e-mail address is whitelisted, then there is no
	 * need to check again, it is whitelisted for all the recipients.
	 * 
	 * Moreover, this will prevent a wrong X-Greylist header display
	 * if the {IP, sender e-mail} address was whitelisted and the
	 * last recipient was also whitelisted. If we would set priv_whitelist
	 * on the last recipient, all recipient would have a X-Greylist
	 * header explaining that they were whitelisted, whereas some
	 * of them would not.
	 */
	if ((priv->priv_whitelist == EXF_ADDR) ||
	    (priv->priv_whitelist == EXF_FROM))
		return SMFIS_CONTINUE;

	/* 
	 * Reload the config file if it has been touched
	 * Restart the sync master thread if nescessary
	 */
	conf_update();
	sync_master_restart();

	/*
	 * Strip spaces from the recipient address
	 */
	strncpy_rmsp(rcpt, *envrcpt, ADDRLEN);
	rcpt[ADDRLEN] = '\0';

	/*
	 * Check if the sender IP, sender e-mail or recipient e-mail
	 * is in the permanent whitelist.
	 */
	if ((priv->priv_whitelist = except_filter(&priv->priv_addr, 
	    priv->priv_from, rcpt, priv->priv_queueid)) != EXF_NONE) {
		priv->priv_elapsed = 0;
		return SMFIS_CONTINUE;
	}

	/* 
	 * Check if the tuple {sender IP, sender e-mail, recipient e-mail}
	 * was autowhitelisted
	 */
	if ((priv->priv_whitelist = autowhite_check(&priv->priv_addr,
	    priv->priv_from, rcpt, priv->priv_queueid)) != EXF_NONE) {
		priv->priv_elapsed = 0;
		return SMFIS_CONTINUE;
	}

	/*
	 * On a multi-recipient message, one message can be whitelisted,
	 * and the next ones be greylisted. The first one would
	 * pass through immediatly (priv->priv_delay = 0) with a 
	 * priv->priv_whitelist = EXF_NONE. This would cause improper
	 * X-Greylist header display in mlfi_eom()
	 *
	 * The fix: if we make it to mlfi_eom() with priv_elapsed = 0, this
	 * means that some recipients were whitelisted. 
	 * We can set priv_whitelist now, because if the message is greylisted
	 * for everyone, it will not go to mlfi_eom(), and priv_whitelist 
	 * will not be used.
	 */
	priv->priv_whitelist = EXF_RCPT;

	/*
	 * Check if the tuple {sender IP, sender e-mail, recipient e-mail}
	 * is in the greylist and if it ca now be accepted. If it is not
	 * in the greylist, it will be added.
	 */
	if (pending_check(&priv->priv_addr, priv->priv_from, 
	    rcpt, &remaining, &priv->priv_elapsed, priv->priv_queueid) != 0) 
		return SMFIS_CONTINUE;

	/*
	 * The message has been added to the greylist and will be delayed.
	 * Log this and report to the client.
	 */
	h = remaining / 3600;
	remaining = remaining % 3600;
	mn = (remaining / 60);
	remaining = remaining % 60;
	s = remaining;

	syslog(LOG_INFO, "%s: addr %s from %s to %s delayed for %02d:%02d:%02d",
	    priv->priv_queueid,
	    inet_ntop(AF_INET, &priv->priv_addr, addrstr, IPADDRLEN),
	    priv->priv_from, *envrcpt, h, mn, s);

	if (quiet) {
		(void)smfi_setreply(ctx, "451", "4.7.1", 
		    "Greylisting in action, please come back later");
	} else {
		snprintf(hdr, HDRLEN, 
		    "Greylisting in action, please come "
		    "back in %02d:%02d:%02d", h, mn, s);
		(void)smfi_setreply(ctx, "451", "4.7.1", hdr);
	}

	return SMFIS_TEMPFAIL;
}

sfsistat
mlfi_eom(ctx)
	SMFICTX *ctx;
{
	struct mlfi_priv *priv;
	char hdr[HDRLEN + 1];
	int h, mn, s;
	char *fqdn = NULL;
	char *ip = NULL;
	char timestr[HDRLEN + 1];
	struct timeval tv;
	char *whystr = NULL;
	char host[ADDRLEN + 1];

	priv = (struct mlfi_priv *) smfi_getpriv(ctx);

	if ((fqdn = smfi_getsymval(ctx, "{j}")) == NULL) {
		syslog(LOG_DEBUG, "smfi_getsymval faild for {j}: %s",
		    strerror(errno));
		gethostname(host, ADDRLEN);
		fqdn = host;
	}

	if ((ip = smfi_getsymval(ctx, "{if_addr}")) == NULL) {
		syslog(LOG_DEBUG, "smfi_getsymval faild for {if_addr}: %s",
		    strerror(errno));
		ip = "0.0.0.0";
	}

	(void)gettimeofday(&tv, NULL);
	strftime(timestr, HDRLEN, 
	    "%a, %d %b %Y %T %Z", localtime((time_t *)&tv.tv_sec));

	if (priv->priv_elapsed == 0) {
		switch (priv->priv_whitelist) {
		case EXF_ADDR:
			whystr = "Sender IP whitelisted";
			break;

		case EXF_FROM:
			whystr = "Sender e-mail whitelisted";
			break;

		case EXF_RCPT:
			whystr = "Recipient e-mail whitelisted";
			break;

		case EXF_AUTO:
			whystr = "IP, sender and recipient auto-whitelisted";
			break;

		default:
			syslog(LOG_ERR, "%s: unexpected priv_whitelist = %d", 	
			    priv->priv_queueid, priv->priv_whitelist);
			whystr = "Internal error";
			break;
		}

		snprintf(hdr, HDRLEN, "%s, not delayed by "
		    "milter-greylist-%s (%s [%s]); %s",
		    whystr, PACKAGE_VERSION, fqdn, ip, timestr);

		smfi_addheader(ctx, HEADERNAME, hdr);

		return SMFIS_CONTINUE;
	}

	h = priv->priv_elapsed / 3600;
	priv->priv_elapsed = priv->priv_elapsed % 3600;
	mn = (priv->priv_elapsed / 60);
	priv->priv_elapsed = priv->priv_elapsed % 60;
	s = priv->priv_elapsed;

	snprintf(hdr, HDRLEN,
	    "Delayed for %02d:%02d:%02d by milter-greylist-%s (%s [%s]); %s", 
	    h, mn, s, PACKAGE_VERSION, fqdn, ip, timestr);
	smfi_addheader(ctx, HEADERNAME, hdr);

	return SMFIS_CONTINUE;
}

sfsistat
mlfi_close(ctx)
	SMFICTX *ctx;
{
	struct mlfi_priv *priv;

	if ((priv = (struct mlfi_priv *) smfi_getpriv(ctx)) != NULL) {
		free(priv);
		smfi_setpriv(ctx, NULL);
	}

	return SMFIS_CONTINUE;
}



int
main(argc, argv)
	int argc;
	char *argv[];
{
	int ch;
	int gotsocket = 0;
	struct passwd *pw = NULL;

	/* Process command line options */
	while ((ch = getopt(argc, argv, "a:vDd:qw:f:hp:Tu:r")) != -1) {
		switch (ch) {
		case 'a':
			if (optarg == NULL) {
				fprintf(stderr, "%s: -a needs an argument\n",
				    argv[0]);
				usage(argv[0]);
			}
			autowhite_validity = (time_t)humanized_atoi(optarg);
			break;

		case 'D':
			dont_fork = 1;
			break;

		case 'q':
			quiet = 1;
			break;

		case 'r':
			printf("milter-greylist-%s %s\n", 
			    PACKAGE_VERSION, BUILD_ENV);
			exit(EX_OK);
			break;

		case 'u': {
			if (geteuid() != 0) {
				fprintf(stderr, "%s: only root can use -u\n", 
				    argv[0]);
				exit(EX_USAGE);
			}

			if ((optarg == NULL) || 
			    ((pw = getpwnam(optarg)) == NULL)) {
				fprintf(stderr, 
				    "%s: -u needs a valid user as argument\n",
				    argv[0]);
				usage(argv[0]);
			}
			break;
		}
			
		case 'v':
			debug = 1;
			break;

		case 'w':
			if ((optarg == NULL) || 
			    ((delay = humanized_atoi(optarg)) == 0)) {
				fprintf(stderr, 
				    "%s: -w needs a positive argument\n",
				    argv[0]);
				usage(argv[0]);
			}
			break;

		case 'f':
			if (optarg == NULL) {
				fprintf(stderr, "%s: -f needs an argument\n",
				    argv[0]);
				usage(argv[0]);
			}
			conffile = optarg;
			break;

		case 'd':
			if (optarg == NULL) {
				fprintf(stderr, "%s: -d needs an argument\n",
				    argv[0]);
				usage(argv[0]);
			}
			dumpfile = optarg;
			break;
				
		case 'p':
			if (optarg == NULL) {
				fprintf(stderr, "%s: -p needs an argument\n",
				    argv[0]);
				usage(argv[0]);
			}
			cleanup_sock(optarg);
			(void)smfi_setconn(optarg);
			gotsocket = 1;
			break;

		case 'T':
			testmode = 1;	
			break;

		case 'h':
		default:
			usage(argv[0]);
			break;
		}
	}
	
	if (gotsocket == 0) {
		fprintf(stderr, "%s: -p is a mandatory option\n",
		    argv[0]);
		usage(argv[0]);
	}

	/*
	 * Drop root privs
	 */
	if (pw != NULL) {
		if ((setuid(pw->pw_uid) != 0) ||
		    (seteuid(pw->pw_uid) != 0)) {
			fprintf(stderr, "%s: cannot change UID: %s\n",
			    argv[0], strerror(errno));
			exit(EX_OSERR);
		}
	}

	/* 
	 * Register our callbacks 
	 */
	if (smfi_register(smfilter) == MI_FAILURE) {
		fprintf(stderr, "%s: smfi_register failed\n", argv[0]);
		exit(EX_UNAVAILABLE);
	}

	/*
	 * Various init
	 */
	if ((except_init() != 0) ||
	    (pending_init() != 0) ||
	    (peer_init() != 0) ||
	    (autowhite_init() != 0) ||
	    (dump_init() != 0)) {
		fprintf(stderr, "%s: list init failed\n", argv[0]);
		exit(EX_SOFTWARE);
	}

	if (dont_fork != 0)
		openlog("milter-greylist", LOG_PERROR, LOG_MAIL);
	else
		openlog("milter-greylist", 0, LOG_MAIL);

	/*
	 * Load config file
	 * We can do this without locking exceptlist, as
	 * normal operation has not started: no other thread
	 * can access the list yet.
	 */
	conf_load();
	
	/*
	 * Reload a saved greylist
	 * No lock needed here either.
	 */
	dump_reload();

	/*
	 * Turn into a daemon
	 */
	if (dont_fork == 0) {

		(void)close(0);
		(void)open("/dev/null", O_RDONLY, 0);
		(void)close(1);
		(void)open("/dev/null", O_RDONLY, 0);
		(void)close(2);
		(void)open("/dev/null", O_RDONLY, 0);

		if (chdir("/") != 0) {
			fprintf(stderr, "%s: cannot chdir to root: %s\n",
			    argv[0], strerror(errno));
			exit(EX_OSERR);
		}

		switch (fork()) {
		case -1:
			fprintf(stderr, "%s: cannot fork: %s\n",
			    argv[0], strerror(errno));
			exit(EX_OSERR);
			break;

		case 0:
			break;

		default:
			exit(EX_OK);	
			break;
		}
	}

	/*
	 * Start the dumper thread
	 */
	dumper_start();

	/*
	 * Run the peer MX greylist sync threads
	 */
	sync_master_restart();
	sync_sender_start();

	/*
	 * Here we go!
	 */
	return smfi_main();
}

void
usage(progname)
	char *progname;
{
	fprintf(stderr, 
	    "usage: %s [-DvqT] [-a autowhite] [-d dumpfile] [-f configfile]\n"
	    "       [-w delay] [-u username] -p socket\n", progname);
	exit(EX_USAGE);
}

void
cleanup_sock(path)
	char *path;
{
	struct stat st;

	/* Does it exists? Get information on it if it does */
	if (stat(path, &st) != 0)
		return;

	/* Is it a socket? */
	if ((st.st_mode & S_IFSOCK) == 0)
		return;

	/* Remove the beast */
	(void)unlink(path);
	return;
}

static char *
strncpy_rmsp(dst, src, len)
	char *dst;
	char *src;
	size_t len;
{
	unsigned int i;

	for (i = 0; src[i] && (i < len); i++) {
		if (isgraph((int)src[i]))
			dst[i] = src[i];
		else
			dst[i] = '_';
	}

	if (i < len)
		dst[i] = '\0';

	return dst;
}

static int
humanized_atoi(str)	/* *str is modified */
	char *str;
{
	unsigned int unit;
	size_t len;

	if ((len = strlen(str)) == 0)
		return 0;

	switch(str[len - 1]) {
	case 's':
		unit = 1;
		break;

	case 'm':
		unit = 60;
		break;

	case 'h':
		unit = 60 * 60;
		break;

	case 'd':
		unit = 24 * 60 * 60;
		break;

	case 'w':
		unit = 7 * 24 * 60 * 60;
		break;

	default:
		return atoi(str);
		break;
	}

	str[len - 1] = '\0';

	return (atoi(str) * unit);
}
